src/inmanta/deploy/state.py

```diff
@@ -210,66 +224,192 @@ def are_dependencies_compliant(self, resource: ResourceIdStr) -> bool:
+        known_blockers_cache: dict[ResourceIdStr, ResourceIdStr] = {}
+        # Cache used by the `update_blocked_status_*()` method to prevent the expensive check on the blocked
+        # status of the requirements of a resource. If the dictionary contains key value pair (X, Y), it
+        # indicates that resource Y is a requirement of resource X that was blocked, last we checked
+        # This cache is a bit complicated in that
+        #  - The unblock check on node A takes us up to n traversals (if A has n requires)
+        #  - We may hit A up to n times
+        #  - The chance of a cache miss is 1/n (i.e. we come from that edge) (roughly)
+        #  - So we reduce from n*n to n time complexity
+        # unclear if it is worth it, hard to tell/test
+
+        def update_blocked_status_to_blocked(resource: ResourceIdStr) -> bool:
+            """
+            Check if the deployment of the given resource is still blocked and mark it as unblocked if it is.
+            check one step over the requires relation, assume all states to hold.
+            if they don't we will pass this node again
 
-        Must be called under the scheduler lock. This method assumes that all the resources of the model version
-        are populated in the resources dictionary.
+            :return: True iff the given resource changes state.
+            """
+            if resource in is_blocked:
+                # Is blocked for a know root
+                return False
 
-        :param resource: The resource that became defined.
-        :param details: The details of resource.
-        """
+            my_state = self.resource_state[resource]
+            if my_state.blocked is BlockedStatus.YES:
+                # The resource is already blocked.
+                return False
 
-        # Cache used by the `update_blocked_status()` method to prevent the expensive check on the blocked
-        # status of the requirements of a resource. If the dictionary contains key value pair (X, Y), it
-        # indicates that resource Y is a requirement of resource X that was blocked, last we checked
-        known_blockers_cache: dict[ResourceIdStr, ResourceIdStr] = {}
+            if my_state.status is ComplianceStatus.UNDEFINED:
+                assert False  # TODO: remove, should not happen
 
+            blocked_dependency: ResourceIdStr | None = None
+            # one blocked requires, evidence we are blocked
+
+            if resource in known_blockers_cache:
+                # First check the blocked status of the cached known blocker for improved performance.
+                known_blocker: ResourceIdStr = known_blockers_cache[resource]
+                if self.resource_state[known_blocker].blocked is BlockedStatus.YES:
+                    blocked_dependency = known_blocker
+                else:
+                    # Cache is out of date. Clear cache item.
+                    del known_blockers_cache[resource]
```

Comment:
I think this is correct: 
1. `known_blockers_cache` contains, for every resource A, a resource B that is seen to be blocking the resource A (by definition a direct dependency of A)
2. if this witness it no longer blocked, we drop the cache entry and continue 

I think this is fairly efficient, as it avoids a reverse cache lookup.

Reply:
Oh wow, it was clearly the end of a long review day. I read this block as `for ... in ...` rather than `if ... in ...` This is exactly what I would expect.